Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1372 @@
+
+To sum it up, here are some pros and cons of Happy in comparison to Megaparsec:
+
+Pros:
+* Parser generators report ambiguous grammars to the user, creating greater trust in the parser's output and saving from debugging headaches.
+* Bottom-up parsers can handle left-recursion, while top-down parsers can't.
+* The performance will be more predictable, as Megaparsec can cause exponential time complexity with the use of `try`. In addition, Happy's "static" generated LALR algorithm has better known complexity constraints.
+* Dealing with operator precedence and associativity requires less labor. In Megaparsec, there are standard tools like [`makeExprParser`](https://hackage.haskell.org/package/parser-combinators-1.3.0/docs/Control-Monad-Combinators-Expr.html#v:makeExprParser) to deal with them, but once you need something that is non-standard, you'll need to write more non-trivial parser rules by hand.
+
+Cons:
+* Less idiomatic, as you need to write Happy grammars instead of Haskell code.
+* Less flexible since you can't change Happy’s algorithm. Meanwhile, parser combinators enable you to easily create your own functions.
+
+## How it works
+
+An in-depth explanation of how an LR(1) parser works is well beyond the scope of this tutorial. For this, we recommend you read our article on [how to Implement an LR(1) Parser](https://serokell.io/blog/how-to-implement-lr1-parser).
+We will, however, introduce a few basic concepts required to understand this article better.
+
+### Terminology
+
+A **language** (formally) is a (likely infinite) set of all grammatically correct sentences.
+
+A **grammar** describes how to construct an arbitrary sentence of the language with simple string substitutions.
+Formally, a grammar consists of terminals, non-terminals, production rules and a designated starting non-terminal.
+
+The process of constructing a sentence using the grammar is called a **derivation**.
+
+Parsing is in some sense a process inverse to derivation, i.e. given the end result (the input string in parsing terms) we recover a derivation that produced it.
+
+A **terminal** is a lexical item of the language. In our case, they will be represented by tokens that were produced by Alex.
+They will represent the leaves of the syntax tree.
+
+A **non-terminal** is a description of how the grammar can group symbols together. It's an arbitrary identifier that's replaced by other terminals and non-terminals (terminals and non-terminals together are called **symbols**) to produce derivations. A sequence that doesn't match any non-terminal is considered a syntax error.
+In the example below, `myNonTerminal` is non-terminal.
+They will represent the internal nodes of the abstract syntax tree.
+
+A **production rule** is a rule describing how to "produce" a string that is syntactically valid according to the grammar. It joins terminals together to form a sentence.
```

Comment:
```suggestion
A **production rule** is a rule describing how to "produce" a string that is syntactically valid according to the grammar. It joins symbols together to form a sentence.
```

Senpai, I saw something that's wrong! 😂

Reply:
This is... Debatable. See, the end result only contains terminals, so the original sentence is not strictly speaking wrong.

Reply:
Hmm, then I guess I didn't understand it well enough, my apologies.